### PR TITLE
Reflection

### PR DESCRIPTION
**Pull request by suVrik**
Done for now.

### CODE REVIEW

File: program/unittest/source/reflection/reflection.cpp [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+    reflection_test_details::no_fields_no_meta<int16*>();
+    reflection_test_details::no_fields_no_meta<int32*>();
+    reflection_test_details::no_fields_no_meta<int64*>();
+    reflection_test_details::no_fields_no_meta<void**>();
+    reflection_test_details::no_fields_no_meta<bool**>();
+    reflection_test_details::no_fields_no_meta<int8**>();
+    reflection_test_details::no_fields_no_meta<int16**>();
+    reflection_test_details::no_fields_no_meta<int32**>();
+    reflection_test_details::no_fields_no_meta<int64**>();
+    // These must be equal to their non-reference bothers.
```
Comment (antkir) [inline program/unittest/source/reflection/reflection.cpp]: >bothers

File: framework/include/kw/reflection/any.h [hunk 2]
```diff
@@ -23,34 +23,57 @@ struct hash;
...
     Any(Any&& original) noexcept;
     ~Any() noexcept;
     Any& operator=(const Any&) = delete;
-    Any& operator              =(Any&& original) noexcept;
+    Any& operator=(Any&& original) noexcept;
 
     /**
-     * Operators equals and less implemented for associative arrays, but can be used for any other reason.
+     * Compare two Any instances using operators provided by their Types.
+     * If these two Any instances have different types, they're considered to be not equal.
```
Comment (antkir) [inline framework/include/kw/reflection/any.h]: unequal

File: framework/include/kw/reflection/type.h [hunk 3]
```diff
@@ -39,39 +40,44 @@ class Type final {
...
+     *
      * If the object doesn't have a hash function, this function will cause an invalid assertion and
-     * return a object's pointer casted to size_t (so it's not completely broken in release build).
+     * return an object's pointer casted to size_t (so it's not completely broken in release build).
      */
     typedef size_t (*hash_t)(const void*) noexcept;
 
     /**
-     * Return an unique Type instance for any unique type T.
+     * Return an unique `Type` instance for any unique type `T`.
```
Comment (antkir) [inline framework/include/kw/reflection/type.h]: a unique

File: program/unittest/source/reflection/reflection.cpp [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+}
+
+template <typename T>
+void fields_meta() {
+    using namespace kw;
+
+    FastName meta_a("c");
+    FastName meta_b("d");
+
+    // Separate reflection adding and using to imitate different scopes
```
Comment (antkir) [inline program/unittest/source/reflection/reflection.cpp]: Hard to understand, "addition and usage" seems a bit better to me.

File: framework/include/kw/reflection/reflection.h [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+     * child meta field.
+     */
+    class Meta;
+
+    /**
+     * Reflection::Field reflects an actual field of an object (i.e. we have an object of type Cat and it has a field
+     * called `name`, reflection of this field is its type, address relative to field's object and name).
+     *
+     * Fields are inherited. If parent reflection already contains a field with specific name, it is overridden by a
+     * child field. But keep in mind, if two parent objects have fields with the same name, it is an assertion failure.
```
Comment (antkir) [inline framework/include/kw/reflection/reflection.h]: I guess "an assertion error will occur" or smth like that makes more sense.
But maybe I'm nitpicking here.

File: framework/include/kw/reflection/reflection.h [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    class Field;
+
+    /**
+     * Reflection::Method reflects an actual method of an object (i.e. if we have an object of type Cat and it has a
+     * method called `meow`, reflection of this method is a `Function` (please take a look at AnyFunction class) and
+     * a name.
+     *
+     * Methods are inherited. If parent reflection already contains a method with specific name, it is overridden by a
+     * child method. But keep in mind, if two parent objects have methods with the same name,
+     * it is an assertion failure.
```
Comment (antkir) [inline framework/include/kw/reflection/reflection.h]: Here as well.

File: framework/include/kw/reflection/reflection.h [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+    // Allow this function to create Meta:
+    friend Meta* reflection_details::add_meta(const FastName&, Any&&) noexcept;
+};
+
+/**
+ * Reflection::Field reflects an actual field of an object (i.e. we have an object of type Cat and it has a field
+ * called `name`, reflection of this field is its type, address relative to field's object and name).
+ *
+ * Fields are inherited. If parent reflection already contains a field with specific name, it is overridden by a
+ * child field. But keep in mind, if two parent objects have fields with the same name, it is an assertion failure.
```
Comment (antkir) [inline framework/include/kw/reflection/reflection.h]: And here.

File: framework/include/kw/reflection/reflection.h [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+};
+
+/**
+ * Reflection::Method reflects an actual method of an object (i.e. if we have an object of type Cat and it has a
+ * method called `meow`, reflection of this method is a `Function` (please take a look at AnyFunction class) and
+ * a name.
+ *
+ * Methods are inherited. If parent reflection already contains a method with specific name, it is overridden by a
+ * child method. But keep in mind, if two parent objects have methods with the same name,
+ * it is an assertion failure.
```
Comment (antkir) [inline framework/include/kw/reflection/reflection.h]: Here too.

**Comment by suVrik** [comment]
[Comment for reflection/ package](https://trello.com/c/KeX1L6Dc/99-comment-for-reflection-package)

**Comment by suVrik** [comment]
[Reflection methods](https://trello.com/c/r8iju80C/100-reflection-methods)

**Comment by suVrik** [comment]
[AnyFunction](https://trello.com/c/CM0yAdF0/101-anyfunction)

**Comment by suVrik** [comment]
[Reflection + Tests](https://trello.com/c/UYDZpHHP/61-reflection-tests)

**Comment by suVrik** [comment]
[Meta + Tests](https://trello.com/c/UYiUo7oL/62-meta-tests)

**Pull request merged**